File: 102.BinaryTreeLevelOrderTraversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        levels = []
        queue = []

        if root == None:
            return levels

        queue.append(root)
        queue.append(None)
        
        level_elements = []
        while len(queue) > 0:
            ele = queue.pop(0)

            if ele == None:
                levels.append(level_elements)
                level_elements = []
                if len(queue) == 0:
                    break
                else:
                    queue.append(None)
            else:
                level_elements.append(ele.val)

                if ele.left != None:
                    queue.append(ele.left)
                if ele.right != None:
                    queue.append(ele.right)
        
        return levels

File: 102.BinaryTreeLevelOrderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {


    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> levels = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();

        if(root == null) {
            return levels;
        }

        queue.add(root);
        queue.add(null);
        List<Integer> levelElements = new ArrayList<>();
        while(!queue.isEmpty()) {
            TreeNode ele = queue.remove();
            if(ele == null) {
                levels.add(levelElements);
                levelElements = new ArrayList<>();
                if (queue.isEmpty()) {
                    break;
                } else {
                    queue.add(null);
                }
            } else {
                levelElements.add(ele.val);

                if(ele.left != null) {
                    queue.add(ele.left);
                }

                if(ele.right != null) {
                    queue.add(ele.right);
                }
            }

        }
        return levels;
    }
}
